src/upgrade.py

```diff
@@ -261,6 +267,29 @@ def _prepare_upgrade_from_legacy(self) -> None:
             )
         self.charm.postgresql.set_up_database()
 
+    def _remove_secrets_old_revisions(self) -> None:
+        """Remove secrets' old revisions."""
+        if self.charm.unit.is_leader():
+            secret = self.charm.model.get_secret(label=f"{PEER}.{self.charm.app.name}.app")
+            secret_id = secret.get_info().id
+            latest_revision = secret.get_info().revision
+            new_env = os.environ.copy()
+            if "JUJU_CONTEXT_ID" in new_env:
+                new_env.pop("JUJU_CONTEXT_ID")
+            for revision in range(1, latest_revision):
+                command = [
+                    self.run_cmd,
+                    self.charm.unit.name,
+                    "--",
+                    "secret-remove",
+                    "--revision",
+                    str(revision),
+                    secret_id,
+                ]
```

Comment:
This also won't work. 
I tested this with just manually running hooks, and juju "queues up" a single revision to be removed. So if you just do:
```sh
juju exec --unit u/0 -- secret-remove --revision 1 $secret; secret-remove --revision 2 $secret
```
Then juju will only remove revision 2.

Reply:
I filed a bug about this behavior:
https://github.com/juju/juju/issues/20805

Reply:
I noticed a different behaviour between the approach you commented about and the one that uses `/usr/bin/juju-exec u/0 -- secret-remove --revision 1 $secret; /usr/bin/juju-exec u/0 -- secret-remove --revision 2 $secret` from inside the unit SSH or ops hook context.

The latter approach "queues up" all the revisions to be removed.